### PR TITLE
Support for aggregate errors

### LINKED ISSUE #86: Serialisation of AggregateError
I'd like to propose a new feature.

[AggregateErrors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError/AggregateError) get serialised as regular errors, displaying only information from the aggregated error.

Including message and stack from the underlying errors, in a similar vain to what has been on https://github.com/pinojs/pino-std-serializers/pull/78 could be beneficial and provide a better picture.

Example

```js
const log = require('pino')()
const inner = new Error('inner')
const aggregate = new AggregatedError([inner], 'aggregated')
log.error(aggregate)
```

Current serialisation
```
{
    "level": 50,
    "time": 1640082406574,
    "pid": 45285,
    "hostname": "hostname1",
    "stack": "AggregateError: aggregated\n    at REPL18:1:7\n    at Script.runInThisContext (node:vm:129:12)\n    at REPLServer.defaultEval (node:repl:562:29)\n    at bound (node:domain:421:15)\n    at REPLServer.runBound [as eval] (node:domain:432:12)\n    at REPLServer.onLine (node:repl:889:10)\n    at REPLServer.emit (node:events:402:35)\n    at REPLServer.emit (node:domain:475:12)\n    at REPLServer.Interface._onLine (node:readline:487:10)\n    at REPLServer.Interface._line (node:readline:864:8)",
    "type": "Error",
    "msg": "aggregated"
}
```

Possible improvement
```
{
    "level": 50,
    "time": 1640082406574,
    "pid": 45285,
    "hostname": "hostname1",
    "stack": "AggregateError: aggregated\n    at REPL18:1:7\n    at Script.runInThisContext (node:vm:129:12)\n    at REPLServer.defaultEval (node:repl:562:29)\n    at bound (node:domain:421:15)\n    at REPLServer.runBound [as eval] (node:domain:432:12)\n    at REPLServer.onLine (node:repl:889:10)\n    at REPLServer.emit (node:events:402:35)\n    at REPLServer.emit (node:domain:475:12)\n    at REPLServer.Interface._onLine (node:readline:487:10)\n    at REPLServer.Interface._line (node:readline:864:8)",
    "type": "Error",
    "msg": "aggregated",
    "aggregateErrors": [{
        "msg": "inner",
        "stack": "Error: inner\n    at REPL18:1:7\n    at Script.runInThisContext (node:vm:129:12)\n    at REPLServer.defaultEval (node:repl:562:29)\n    at bound (node:domain:421:15)\n    at REPLServer.runBound [as eval] (node:domain:432:12)\n    at REPLServer.onLine (node:repl:889:10)\n    at REPLServer.emit (node:events:402:35)\n    at REPLServer.emit (node:domain:475:12)\n    at REPLServer.Interface._onLine (node:readline:487:10)\n    at REPLServer.Interface._line (node:readline:864:8)"
    }]
}

### PR DESCRIPTION
**Pull request by sameer-coder**
Example

```js
const log = require('pino')()
const foo = new Error('foo')
const bar = new Error('bar')
const aggregate = new AggregateError([foo, bar], 'aggregated message')
log.error(aggregate)
```

Produces the following serialized output:
```json
{
  "level": 50,
  "time": 1641313649352,
  "pid": 1788,
  "hostname": "SAMEER",
  "err": {
    "type": "AggregateError",
    "message": "aggregated message",
    "stack": "AggregateError: aggregated message\n at Object.<anonymous> (pino-std-serializers\\example.js:12:19)\n at Module._compile (node:internal/modules/cjs/loader:1108:14)\n at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)\n at Module.load (node:internal/modules/cjs/loader:988:32)\n at Function.Module._load (node:internal/modules/cjs/loader:828:14)\n at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)\n at node:internal/main/run_main_module:17:47",
    "aggregateErrors": [
      {
        "type": "Error",
        "message": "foo",
        "stack": "Error: foo\n at Object.<anonymous> (pino-std-serializers\\example.js:10:15)\n at Module._compile (node:internal/modules/cjs/loader:1108:14)\n at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)\n at Module.load (node:internal/modules/cjs/loader:988:32)\n at Function.Module._load (node:internal/modules/cjs/loader:828:14)\n at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)\n at node:internal/main/run_main_module:17:47"
      },
      {
        "type": "Error",
        "message": "bar",
        "stack": "Error: bar\n at Object.<anonymous> (pino-std-serializers\\example.js:11:15)\n at Module._compile (node:internal/modules/cjs/loader:1108:14)\n at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)\n at Module.load (node:internal/modules/cjs/loader:988:32)\n at Function.Module._load (node:internal/modules/cjs/loader:828:14)\n at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)\n at node:internal/main/run_main_module:17:47"
      }
    ]
  },
  "msg": "aggregated message"
}
```

*Note: AggregateError is only supported by node 15+*

Closes #86

### CODE REVIEW

File: test/err.test.js [hunk 1]
```diff
@@ -1,9 +1,29 @@
 'use strict'
 
-const test = require('tap').test
+const { test, skip } = require('tap')
```
Comment (mcollina) [inline test/err.test.js:3]: why `skip`?

Comment (sameer-coder) [reply test/err.test.js:3]: I am skipping the test if node version is < 15.
Used it [here ](https://github.com/pinojs/pino-std-serializers/pull/87/files#diff-a6ff895a4ce70729791324c8118f43cffada5a9e9cfe42ea1df4a18707f2c430R9)

File: lib/err.js [hunk 2]
```diff
@@ -4,6 +4,7 @@ module.exports = errSerializer
 
 const { messageWithCauses, stackWithCauses } = require('./err-helpers')
 
+const major = Number(process.versions.node.split('.')[0])
```
Comment (mcollina) [inline lib/err.js:7]: instead of relying on a version of Node.js, please use an existence check on `global.AggregateError`

File: test/err.test.js [hunk 3]
```diff
@@ -1,9 +1,29 @@
...
 const serializer = require('../lib/err')
 const wrapErrorSerializer = require('../').wrapErrorSerializer
 
+test('serializes aggregate errors', function (t) {
+  const major = Number(process.versions.node.split('.')[0])
+  if (major < 15) return skip()
```
Comment (mcollina) [inline test/err.test.js:9]: Please avoid the `if` and use this pattern instead:

https://github.com/pinojs/pino/blob/df49265d889788c7372e9bd9348261df712d2198/test/exit.test.js#L42

File: package.json [hunk 4]
```diff
@@ -8,8 +8,8 @@
   "scripts": {
     "lint": "standard | snazzy",
     "lint-ci": "standard",
-    "test": "tap --no-cov 'test/**/*.test.js'",
-    "test-ci": "tap --cov --no-check-coverage --coverage-report=text 'test/**/*.test.js'",
+    "test": "tap --no-cov test/**/*.test.js",
+    "test-ci": "tap --cov --no-check-coverage --coverage-report=text test/**/*.test.js",
```
Comment (mcollina) [inline package.json:12]: Why those changes?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (sameer-coder) [reply package.json:12]: I have only removed the quotes from the path. The `test` script wouldn't run without it

File: lib/err.js (hunk 2 above)
Comment (sameer-coder) [reply lib/err.js]: yep. That's a better way!

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/err.js [hunk 5]
```diff
@@ -50,6 +56,10 @@ function errSerializer (err) {
     : err.name
   _err.message = messageWithCauses(err)
   _err.stack = stackWithCauses(err)
+  // eslint-disable-next-line no-undef
+  if (hasAggregateError && err instanceof AggregateError && err.errors && Array.isArray(err.errors)) {
+    _err.aggregateErrors = err.errors.map(errSerializer)
```
Comment (voxpelli) [inline lib/err.js:61]: This bypasses the `if (!val.hasOwnProperty(seen)) {` check and thus is susceptible to infinite recursion if an aggregate error includes itself in the list of errors.

We should add a `prevents infinite aggregate recursion` test + either add a check right here or in the beginning of `errSerializer` to ensure that infinite is aborted.

**Review by voxpelli — CHANGES_REQUESTED** [review]
Some comments, thoughts, questions – but most importantly: A note that this is susceptible to infinite recursion. Only the latter is why I set "Request changes" :)

File: lib/err.js (hunk 5 above)
Comment (sameer-coder) [reply lib/err.js:61]: @voxpelli I agree that infinite recursion is possible if the aggregate error includes itself in the list of errors but can you provide a scenario where this could happen? I think creating a new aggregate error requires you to provide an iterable of already existing errors so the list of errors cannot contain a reference to the aggregate error itself.

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (sameer-coder) [reply package.json:12]: I think the it is still performed by tap. I check this on windows and it works fine

[the remark this answers is not in the document]
Comment (mcollina) [reply package.json:12]: I would still prefer these to be reverted and fixed in another PR if they __must__ be changed.

File: lib/err.js (hunk 5 above)
Comment (mcollina) [reply lib/err.js]: @sameer-coder I'm kind of lost on why it's impossible to write a unit test that cover this case. Can you add one anyway?

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (jsumners) [reply package.json:12]: _If_ they are going to be changed, then they should be moved to `.taprc` by:

```yaml
files:
  - 'test/**/*.test.js'
```

Otherwise, the globbing will be interpreted by the shell as @voxpelli stated.

File: lib/err.js [hunk 6]
```diff
@@ -4,6 +4,7 @@ module.exports = errSerializer
 
 const { messageWithCauses, stackWithCauses } = require('./err-helpers')
 
+const hasAggregateError = !!global.AggregateError
```
Comment (jsumners) [inline lib/err.js:7]: Can we change this check to something easier to read? Double negates are difficult.

File: lib/err.js [hunk 7]
```diff
@@ -50,6 +56,10 @@ function errSerializer (err) {
     : err.name
   _err.message = messageWithCauses(err)
   _err.stack = stackWithCauses(err)
+  // eslint-disable-next-line no-undef
```
Comment (jsumners) [inline lib/err.js:59]: What is this comment for? Can we resolve it another way?

**Review by jsumners — CHANGES_REQUESTED** [review]

File: lib/err.js (hunk 6 above)
Comment (voxpelli) [reply lib/err.js:7]: @jsumners Which one do you like better?

1. `global.AggregateError !== undefined`
2. `typeof global.AggregateError === 'object'`
3. `Boolean(global.AggregateError)`

(I myself like `!!`)

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (sameer-coder) [reply package.json:12]: I will revert this change

File: lib/err.js (hunk 7 above)
Comment (voxpelli) [reply lib/err.js:59]: If changed to refer to `global.AggregateError`, like in the check earlier, then the comment can go away

File: lib/err.js (hunk 6 above)
Comment (jsumners) [reply lib/err.js:7]: `global.AggregateError !== undefined`

File: lib/err.js (hunk 7 above)
Comment (voxpelli) [reply lib/err.js:59]: Side note: we released a prerelease of an ESLint 8 based version of `eslint-config-standard` a few days ago, when that one moves through to a new major of `standard`, then this will be able to get a more proper solution. See https://github.com/eslint/eslintrc/pull/28 and https://github.com/standard/eslint-config-standard/issues/208

File: test/err.test.js [hunk 8]
```diff
@@ -177,3 +177,36 @@ test('can wrap err serializers', function (t) {
...
+  t.match(serialized.aggregateErrors[0].stack, /^Error: foo/)
+  t.match(serialized.aggregateErrors[1].stack, /^Error: bar/)
+  t.match(serialized.stack, /err\.test\.js:/)
+})
+
+test('avoid infinite recursion in aggregate errors', { skip: !global.AggregateError }, function (t) {
+  t.plan(7)
+  const foo = new Error('foo')
+  const agg1 = new AggregateError([foo], 'aggregated message') // eslint-disable-line no-undef
+  agg1.inner = agg1
```
Comment (sameer-coder) [inline test/err.test.js:202]: Not really sure if this test is written correctly

Comment (jsumners) [reply test/err.test.js:202]: Wouldn't this be `agg1.errors.push(agg1)`? Although, I agree with you. I'd like to have a clear explanation of when we think this would actually happen, @voxpelli. An `AggregateError` must be initialized with an array of existing errors. Thus, it cannot be initialized with itself in the aggregate. Anyone appending to the aggregate to its `errors` list seems a Bad Idea, and a scenario we should just classify as unsupported.

http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError

Also, the spec requires at least the array of errors be present when creating a new aggregate -- https://tc39.es/ecma262/multipage/fundamental-objects.html#sec-aggregate-error-constructor

Comment (sameer-coder) [reply test/err.test.js:202]: @voxpelli Did you get a chance to look at this again?

Comment (voxpelli) [reply test/err.test.js:202]: > Anyone appending to the aggregate to its errors list seems a Bad Idea, and a scenario we should just classify as unsupported.

I agree, but since the consequences are catastrophic, I think we should still prevent it.

Same would apply for Error Causes.

Comment (jsumners) [reply test/err.test.js:202]: @mcollina do you have any thoughts on this? @davidmarkclements ?

Comment (mcollina) [reply test/err.test.js:202]: I don't think cycles are a problem. It will be catastrophic but also very easy to debug and fix.
They will need to modify the `.errors` property manually.

To prevernt this, I would just wrap the function in a try/catch

Comment (jsumners) [reply test/err.test.js:202]: So, to be clear, we are agreeing to remove the edge case detection in this code?

Comment (voxpelli) [reply test/err.test.js:202]: And if someone jumps on eg. huntr.dev and reports this as a potential DDoS vector, then we'll simply say that its not valid?

I'm okay with that 👍

Comment (jsumners) [reply test/err.test.js:202]: Yes. As stated, it's unsupported.

Comment (sameer-coder) [reply test/err.test.js]: @jsumners @voxpelli  I have updated the code
